core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -924,6 +923,9 @@ private[spark] class TaskSetManager(
         s" be re-executed (either because the task failed with a shuffle data fetch failure," +
         s" so the previous stage needs to be re-run, or because a different copy of the task" +
         s" has already succeeded).")
+    } else if (sched.stageIdToFinishedPartitions.getOrElse(
+      stageId, new HashSet[Int]).contains(tasks(index).partitionId)) {
```

Comment:
this may create empty hash set unnecessarily. how about 
`sched.stageIdToFinishedPartitions.get(stageId).exists(partitions => partitions.contains(tasks(index).partitionId))`

Reply:
Makes sense, have changed it.